drivers/wifi/at_parser.c

```diff
@@ -0,0 +1,1181 @@
+	{{(uint8_t *)"+GSLP", 5}, AT_SET_OP},
+	{{(uint8_t *)"+GMR", 4}, AT_EXECUTE_OP},
+	{{(uint8_t *)"+CWMODE", 7}, AT_QUERY_OP | AT_SET_OP | AT_TEST_OP},
+	{{(uint8_t *)"+CWJAP", 6}, AT_QUERY_OP | AT_SET_OP},
+	{{(uint8_t *)"+CWLAP", 6}, AT_EXECUTE_OP},
+	{{(uint8_t *)"+CWQAP", 6}, AT_EXECUTE_OP},
+	{{(uint8_t *)"+CWSAP", 6}, AT_QUERY_OP | AT_SET_OP},
+	{{(uint8_t *)"+CWLIF", 6}, AT_EXECUTE_OP},
+	{{(uint8_t *)"+CIPSTATUS", 10}, AT_EXECUTE_OP},
+	{{(uint8_t *)"+CIPSTART", 9}, AT_TEST_OP | AT_SET_OP},
+	{{(uint8_t *)"+CIPSEND", 8}, AT_SET_OP},
+	{{(uint8_t *)"+CIPCLOSE", 9}, AT_EXECUTE_OP | AT_SET_OP},
+	{{(uint8_t *)"+CIFSR", 6}, AT_EXECUTE_OP},
+	{{(uint8_t *)"+CIPMUX", 7}, AT_QUERY_OP | AT_SET_OP},
+	{{(uint8_t *)"+CIPSERVER", 10}, AT_SET_OP},
+	{{(uint8_t *)"+CIPMODE", 8}, AT_QUERY_OP | AT_SET_OP},
+	{{(uint8_t *)"+CIPSTO", 7}, AT_QUERY_OP | AT_SET_OP},
+	{{(uint8_t *)"+PING", 5}, AT_SET_OP}
+};
+
+/* Structure storing a circular buffer for a connection */
+struct at_circular_buff {
+	/* Size of the buffer */
+	uint32_t	size;
+	/* Address to the buffer */
+	uint8_t		*buff;
+	/* Read index */
+	uint32_t	ri;
+	/* Write index */
+	uint32_t	wi;
+	/* Write index */
+	uint32_t	count;
+	/* Last read len */
+	uint32_t	last_len;
+	/* Set if overflow */
+	bool		is_overflow;
+	/* Used for polling in blocking read */
+	bool		new_data;
+};
+
+/* Structure storing a connection status */
+struct connection_desc {
+	/* Connection buffer */
+	volatile struct at_circular_buff	cbuff;
+	/* Pending bytes for reading */
+	uint32_t				to_read;
+	/* True if connection established */
+	bool					active;
+	/* Type of connection */
+	enum socket_type			type;
+};
+
+/* Structure storing the status of the parser */
+struct at_desc {
+	// Uart related fields
+
+	/* Uart descriptor */
+	struct uart_desc	*uart_desc;
+	/* Irq descriptor */
+	struct irq_ctrl_desc	*irq_desc;
+	/* Uart irq id */
+	uint32_t		uart_irq_id;
+
+
+	//Connection related fields
+
+	/* Structures storing connections status */
+	struct connection_desc	conn[MAX_CONNECTIONS];
+	/* Id of active connection or -1 if non is active */
+	int32_t			current_conn;
+	/* Stores if running as single or multiple connection mode */
+	bool			multiple_conections;
+
+	//Data fields
+
+	/* Static allocated buffers */
+	struct {
+		uint8_t	result_buff[RESULT_BUFF_LEN];
+		uint8_t	cmd_buff[CMD_BUFF_LEN];
+	} 			buffers;
+	/* Stores the output for the application */
+	struct circular_buffer	*cb_response;
+	/* Stores data received from the module */
+	volatile struct at_buff	result;
+	/* Buffer to build the command */
+	struct at_buff		cmd;
+	/* Buffer to read one char */
+	uint8_t			read_ch;
+
+
+	// Control fields
+
+	/* State of the callback */
+	volatile enum {
+		/* Normal mode. Read each char and interpret the result */
+		READING_RESPONSES,
+		/* When an +IPD is received, callback enter in this mode */
+		READING_PAYLOAD
+	}			callback_operation;
+	/* Used for AT_SEND */
+	volatile bool		waiting_send;
```

Comment:
I don't think it's a good idea to have a combination of a state machine and some other variables around it that are also some soft of state machines in themselves.
Consider creating more states in `callback_operation` in such a way as to eliminate the need for the `waiting_send` boolean.
This way there is only 1 clear state machine, easier to understand, code, debug.